### PR TITLE
Adding metrics events for blockaid

### PR DESCRIPTION
**Pull request by jpuri**
Adding controller metrics events for blockaid

### CODE REVIEW

**Review by segun — DISMISSED** [review]

**Review by blackdevelopa — DISMISSED** [review]

File: app/scripts/controllers/transactions/index.js
Comment (digiwand) [inline app/scripts/controllers/transactions/index.js:1608]: [nit] what if we pass `securityAlertResponse` in options param since it won't be needed for all _createTransaction calls?

File: app/scripts/controllers/transactions/index.js
Comment (digiwand) [inline app/scripts/controllers/transactions/index.js:2516]: [nit] would be nice to create a constant for `not_applicable` to avoid developer typos. 

we can add these to `BlockaidReason` and `BlockaidResultType` located in `shared/constants/security-provider.ts`. When we add I suggest adding a comment to separate the PPOM produced values and our own locally defined variables e.g. maybe

```
/**
 * Locally defined
 */
 
 notApplicable: 'not_applicable',
```

File: app/scripts/controllers/transactions/index.js
Comment (digiwand) [inline app/scripts/controllers/transactions/index.js:2336]: Id like to discuss changing the schema here

1. securityAlertResponse seems to be a synonym for our existing securityProviderResponse. As we may introduce even more security providers this is not a scalable approach and may confuse the reader. From the reader's level, the name is lacking context to which response this may be referring to, OpenSea vs. Blockaid direct response.

2. for scaling, earlier this month @brad-decker provided thoughts on the tx data schema and suggested a new tx object to hold this data:
[Suggestion from @brad-decker](https://consensys.slack.com/archives/C04C0F6SSJH/p1689105741047959?thread_ts=1688552808.747329&channel=C04C0F6SSJH&message_ts=1689105741.047959) 
> txData.integrationData 

I think we could also consider data storing like:
- txData.integrationData.securityBlockaidResponse
- txData.integrationData.securityProviderResponses.blockaid
- txData. securityProviderResponses.blockaid

Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:2288]: @digiwand : `securityProviderResponses` is deprecated and will go away as we are adding blockaid. `securityAlertResponse` will be generic field to hold security response from any security provider. As I understand at one time only one security provider will be used.

File: shared/constants/security-provider.ts
Comment (jpuri) [reply shared/constants/security-provider.ts:58]: Done

File: app/scripts/controllers/transactions/index.js
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js]: Fixed

**Review by segun — DISMISSED** [review]

File: app/scripts/controllers/transactions/index.js
Comment (digiwand) [inline app/scripts/controllers/transactions/index.js:333]: why add `/* eslint-disable */` instead of adding the `@param` definition for `opts.securityAlertResponse`? 

I am thinking when `securityAlertResponse` isn't included for non-Blockaid builds, and with the jsdoc comment declared, this shouldn't affect the builds

File: app/scripts/controllers/transactions/index.js
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:2301]: @digiwand `securityAlertResponse` is used for security feature for metrics and thus I would prefer to use it consistently.

I am not sure if it will actually be possible for 2 security models to be active at the same time.

File: app/scripts/controllers/transactions/index.js
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:333]: Code fencing is creating issue with it.

File: app/scripts/controllers/transactions/index.js
Comment (cryptotavares) [inline app/scripts/controllers/transactions/index.js:329]: You need to re-add this line for linting. If I am not mistaken we have the `jsdoc/newline-after-description` rule enabled

File: app/scripts/controllers/transactions/index.js
Comment (cryptotavares) [reply app/scripts/controllers/transactions/index.js:333]: can you just disable the linting rules conflicting here? And re-enable them right after the jsdoc?
something like this:
```js
  /* eslint-disable jsdoc/require-param, jsdoc/check-param-names */
  /**
   * Add a new unapproved transaction
   * @param {object} txParams - Standard parameters for an Ethereum transaction
   * @param {object} opts - Options
   * @param {string} opts.actionId - Unique ID to prevent duplicate requests
   * @param {string} opts.method - RPC method that requested the transaction
   * @param {string} opts.origin - Origin of the transaction request, such as the hostname of a dApp
   * @param {boolean} opts.requireApproval - Whether the transaction requires approval by the user
   * @param {object[]} opts.sendFlowHistory - Associated history to store with the transaction
   * @param {object} opts.swaps - Options specific to swap transactions
   * @param {boolean} opts.swaps.hasApproveTx - Whether this transaction required an approval transaction
   * @param {boolean} opts.swaps.meta - Additional metadata to store for the transaction
   * @param {TransactionType} opts.type - Type of transaction to add, such as 'cancel' or 'swap'
   * @returns {Promise<{transactionMeta: TransactionMeta, result: Promise<string>}>} An object containing the transaction metadata, and a promise that resolves to the transaction hash after being submitted to the network
   */
  /* eslint-enable jsdoc/require-param, jsdoc/check-param-names */
```

Comment (cryptotavares) [reply app/scripts/controllers/transactions/index.js:333]: @digiwand it fails due to code fencing... we run linting on fenced files when are not doing a dev build. Meaning that if we build with blockaid enabled (for example in flask) we needed the jsdoc param. But if you build for a build type without blockaid feature, then we get an linting error (jsdoc param does not exist). So the only way is to disable linting rules.

Comment (digiwand) [reply app/scripts/controllers/transactions/index.js]: oh didn't know we are running linting on the fenced files. makes sense now. ty for the info!

File: app/scripts/metamask-controller.js
Comment (cryptotavares) [inline app/scripts/metamask-controller.js:3579]: Won't have much of an impact, but this could also be code-fenced

File: app/scripts/controllers/transactions/index.js
Comment (digiwand) [reply app/scripts/controllers/transactions/index.js:2254]: I don't mind metrics differing as much as the naming relationship between server to client code. I think my main concern is that part of the Blockaid code uses "security provider" and the other part uses "security alert".  

I think I am seeing it another way now... I see securityAlertEnabled is already used in the PPOM repo. I am now also seeing a Figma proposal where "Security Alert" may appear in settings ([Figma](https://www.figma.com/file/Q69vHaA7Or7Rxz4vdzjVHC/Support-for-tx-security-providers-(MVP)?type=design&node-id=4806-25387&mode=design&t=Q6cGbAMaZ0ixbStk-0)). Based on this, I think it's okay to use `securityAlertResponse`... I still have concerns about scaling with this schema as the designs imply we may support multiple providers.

Comment (digiwand) [reply app/scripts/controllers/transactions/index.js:2254]: thank you for the discussion and this is no longer a blocker for me

File: shared/constants/security-provider.ts
Comment (digiwand) [inline shared/constants/security-provider.ts:60]: request to update the value to match Pascal casing of other values
`not_applicable` → `NotApplicable`

also mentioned in other PR: https://github.com/MetaMask/metamask-extension/pull/20125#discussion_r1278091150

Comment (jpuri) [reply shared/constants/security-provider.ts:60]: This is suggested by product team.

File: app/scripts/metamask-controller.js
Comment (jpuri) [reply app/scripts/metamask-controller.js:3579]: Done

File: shared/constants/security-provider.ts
Comment (digiwand) [reply shared/constants/security-provider.ts:60]: I would prefer to keep the same format in the object and then reformat / serialize at the metric call(s)... maybe we can bring in another opinion. @blackdevelopa, @segun, @cryptotavares  since you 3 are here also. thoughts?

Comment (jpuri) [reply shared/constants/security-provider.ts:60]: I updated the value.

**Review by digiwand — APPROVED** [review]

**Review by cryptotavares — APPROVED** [review]

**Pull request merged**
